l2-contracts/contracts/ValidatorRegistry.sol

```diff
@@ -0,0 +1,177 @@
+        // Check if validator with the same public key already exists.
+        for (uint256 i = 0; i < validatorIndexes.length; i++) {
+            uint256 idx = validatorIndexes[i];
+            require(!compareBytes(validators[idx].pubKey, pubKey), "pubKey already exists");
+        }
+
+        verifyPoP(pubKey, pop);
+
+        validatorSequence++;
+        uint256 idx = validatorSequence;
+        validators[idx] = Validator(weight, pubKey, pop, false, 0);
+        validatorIndexes.push(idx);
+        return idx;
+    }
+
+    // Removes a validator. Should fail if the validator is still active or
+    // the inactivity delay has not passed.
+    function remove(uint256 idx) public onlyOwner {
+        verifyExists(idx);
+        require(validators[idx].isInactive, "Validator is still active");
+        require(epoch >= validators[idx].inactiveSince + INACTIVITY_DELAY,
+            "Validator's inactivity delay has not passed"
+        );
+
+        // Remove from mapping.
+        delete validators[idx];
+
+        // Remove from array by swapping the last element (gas-efficient, not preserving order).
+        uint256 offset = idxOffset(idx);
+        validatorIndexes[offset] = validatorIndexes[validatorIndexes.length - 1];
+        validatorIndexes.pop();
+    }
+
+    // Inactivates a validator.
+    function inactivate(uint256 idx) public onlyOwner {
+        verifyExists(idx);
+        validators[idx].isInactive = true;
+        validators[idx].inactiveSince = epoch;
+    }
+
+    // Activates a validator.
+    function activate(uint256 idx) public onlyOwner {
+        verifyExists(idx);
+        validators[idx].isInactive = false;
+    }
+
+    // Changes the weight.
+    function changeWeight(uint256 idx, uint256 weight) public onlyOwner {
+        verifyExists(idx);
+        validators[idx].weight = weight;
+    }
+
+    // Changes the public key and PoP.
+    function changePublicKey(
+        uint256 idx,
+        bytes calldata pubKey,
+        bytes calldata pop
+    ) public onlyOwner {
+        verifyExists(idx);
+        verifyPoP(pubKey, pop);
+        validators[idx].pubKey = pubKey;
+        validators[idx].pop = pop;
+    }
+
+    // Creates a new committee list which will become the next committee list.
+    function setNextCommittee() external onlyOwner {
+        epoch += 1;
+
+        // Replace current committee with next committee
+        delete committee;
+        for (uint256 i = 0; i < nextCommittee.length; i++) {
+            committee.push(nextCommittee[i]);
+        }
+
+        // Populate `nextCommittee` based on active validators
+        delete nextCommittee;
+        for (uint256 i = 0; i < validatorIndexes.length; i++) {
+            uint256 idx = validatorIndexes[i];
+            Validator memory validator = validators[idx];
+            if (!validator.isInactive) {
+                nextCommittee.push(CommitteeValidator(idx, validator.weight, validator.pubKey));
```

Comment:
Here again I'd ask if there's a use case for deactivating all validators and leaving the next committee empty. IIUC the flow to L1 would be that the previous committee has to sign over to the next committee, but once that's empty it's probably over for the L2 as there won't be any more handovers. CometBFT in particular just threw up its hands when the committee was empty, rightfully so, but I can imagine it being some kind of shutdown-the-lights procedure.

Reply:
If you can inactivate all validators, then you have control over all validators. You can also just refuse to produce blocks.

Reply:
I'm not asking from the perspective of validators refusing to participate. Of course there is nothing anybody can do about that. I'm only asking whether there is some kind of chaining between committees, where for example if we have committee N, then empty committee N+1, then non-empty committee N+2, then to move from N to N+2 requires a signature from N+1, which would mean it cannot happen and we're stuck because the owner made a mistake and deactivated everyone. 

Like if the owner accidentally removes all validators, consensus cannot produce an L2 block because nobody can sign it, so how can anyone reactivate it if it lives on L2?

Reply:
I see. If the owner removes all validators and that gets all the way to L1, then we can only recover via hard fork (or if we have some inactivity recovery mechanism on L1).
Let's not worry about this now. We can easily add guard rails to the contract later (although we can never prevent the owner from doing a series of mistakes that halts the chain).